### PR TITLE
Feature.lambdas

### PR DESCRIPTION
**Pull request by jneen**
Lambdas and accessor syntax!

Adds lambda syntax as follows:

```
?x => x + 1
?x ?y => x + y
```

as well as three functions that use them:

```
each(list, ?element => ...) // equivalent to ruby .map
filter(list, ?element => true or false) // equivalent to ruby .filter or .select
roll(init, list, ?accum ?element => new_accum) // equivalent to inject or fold
```

This also introduces accessor syntax for structs and certain declared types:

```
{ a: 1 }.a // 1
filter(field:event_times, ?et => et.type = 'setup')
```

In order to register a declared type as having struct accessors, we must use the `structable` declaration in the declared type:

``` ruby
  Dentaku::Type.declare(:event_type) do
    structable(
      type: ':string',
      start_time: ':numeric',
      end_time: ':numeric',
    )
  end
```

### CODE REVIEW

**Comment by hale** [comment]
Nice, @jneen! 

As an exercise, I'd like to see what this would look like for a real example. I think it will be instructive for anything else we need and to show how this will work once it's handed to config.

For example, in https://github.com/opencounter/opencounter/issues/3202 Jeff describes a fee structure that is based on the number of hours the event lasts per day. The fee is charged as follows: each hour the event takes place costs $175, with a minimum 4h charge per day. In other words, there is a minimum charge of $700 per day for the first four hours, and then any hour above that is an additional $175 per hour. 

Would a function that calculates that fee look something like this?

```
sum(map(hours_per_day(field:event_times), ?cost, ?hours => ceiling(700, 175 * hours))
```

(Assumes a function `hours_per_day` that groups the event periods per day and sums their hours, since periods can span multiple days....not sure if that's the way to go but just an example.)

2. What is the type of the accumulator for `roll`, is it always an integer? What is it's initial value?

File: lib/dentaku/ast/arithmetic.rb [hunk 1]
```diff
@@ -28,30 +28,18 @@ class Addition < Arithmetic
       def operator
         :+
       end
-
-      def self.precedence
```
Comment (joshuabates) [inline lib/dentaku/ast/arithmetic.rb:32]: did we stop needing this after the parser rewrite?

File: lib/dentaku/ast/case.rb [hunk 2]
```diff
@@ -1,16 +1,6 @@
-require_relative 'case/case_conditional'
-require_relative 'case/case_when'
-require_relative 'case/case_then'
-require_relative 'case/case_switch_variable'
-require_relative 'case/case_else'
```
Comment (joshuabates) [inline lib/dentaku/ast/case.rb:5]: How are we able to delete all these case child nodes?

**Review by joshuabates — APPROVED** [review]
👍

File: lib/dentaku/ast/case.rb (hunk 2 above)
Comment (jneen) [reply lib/dentaku/ast/case.rb:5]: Leftover from the parser

File: lib/dentaku/ast/arithmetic.rb (hunk 1 above)
Comment (jneen) [reply lib/dentaku/ast/arithmetic.rb:32]: Yep! Precedence is now hardcoded in the parser, which is honestly where it should live.

**Comment by jneen** [comment]
> Nice, @jneen!
> 
> As an exercise, I'd like to see what this would look like for a real example. I think it will be instructive for anything else we need and to show how this will work once it's handed to config.
> 
> For example, in [opencounter/opencounter#3202](https://github.com/opencounter/opencounter/issues/3202) Jeff describes a fee structure that is based on the number of hours the event lasts per day. The fee is charged as follows: each hour the event takes place costs $175, with a minimum 4h charge per day. In other words, there is a minimum charge of $700 per day for the first four hours, and then any hour above that is an additional $175 per hour.



> Would a function that calculates that fee look something like this?
> 
> ```
> sum(map(hours_per_day(field:event_times), ?cost, ?hours => ceiling(700, 175 * hours))
> ```

Something similar to this!

```
sum(each(seconds_by_day_for_type(field:event_times, 'event'), ?seconds => 175 * clamp_below(4, floor(seconds/(60*60)))))
```

> (Assumes a function `hours_per_day` that groups the event periods per day and sums their hours, since periods can span multiple days....not sure if that's the way to go but just an example.)

This could now be theoretically written in Dentaku as well.

>     2. What is the type of the accumulator for `roll`, is it always an integer? What is it's initial value?

Fixed the example, the initial value should be passed in. The type can be any type, just must be the same as the initial value and the result.

**Comment by jneen** [comment]
If I'm honest, I want to start by introducing `filter`, since it's imo the most intuitive use of lambda. I want to hold off on explaining `roll` until we have an actual use case - everything I've seen so far should be able to be handled by `filter` and `each`, with the occasionall `sum` (which is secretly a fold, but much simpler to understand.)

**Review by hale — APPROVED** [review]

**Pull request merged**